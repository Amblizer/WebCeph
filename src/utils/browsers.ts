import bowser from 'bowser';

const ChromeIcon: string = require('url!./icons/chrome.svg');
const FirefoxIcon: string = require('url!./icons/firefox.svg');
const EdgeIcon: string = require('url!./icons/edge.svg');
const OperaIcon: string = require('url!./icons/opera.svg');
const SafariIcon: string = require('url!./icons/safari-ios.svg');

interface BrowserDetails {
  name: string;
  downloadUrl: string;
  icon: string;
  isApplicable?(): boolean;
}

const _current = bowser;

export const details = {
  Chrome: {
    name: 'Chrome',
    downloadUrl: 'https://google.com/chrome',
    icon: ChromeIcon,
  },
  Firefox: {
    name: 'Firefox',
    downloadUrl: 'https://getfirefox.com/',
    icon: FirefoxIcon,
  },
  Opera: {
    name: 'Opera',
    downloadUrl: 'https://opera.com/',
    icon: OperaIcon,
  },
  'Microsoft Edge': {
    name: 'Edge',
    downloadUrl: 'https://www.microsoft.com/windows/microsoft-edge',
    icon: EdgeIcon,
    isApplicable() {
      return _current.windows && _current.osversion === '10';
    }
  },
  Safari: {
    name: 'Safari',
    downloadUrl: 'https://www.apple.com/safari/',
    icon: SafariIcon,
    isApplicable() {
      return _current.mac;
    }
  },
} as { [id: string]: BrowserDetails };


export const currentBrowser: Browser = {
  id: _current.name as BrowserId,
  name: details[_current.name] ? details[_current.name].name : 'Unknown',
  downloadUrl: details[_current.name] ? details[_current.name].downloadUrl : '',
  version: _current.version,
}

export default details;