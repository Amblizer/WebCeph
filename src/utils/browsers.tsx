import React from 'react';
import bowser from 'bowser';

const ChromeIcon = require('./icons/chrome.svg').default;
const FirefoxIcon = require('./icons/firefox.svg').default;
const EdgeIcon = require('./icons/edge.svg').default;
const OperaIcon = require('./icons/opera.svg').default;
const SafariIcon = require('./icons/safari-ios.svg').default;

interface BrowserDetails {
  name: string;
  downloadUrl: string;
  icon: JSX.Element;
  isApplicable?(): boolean;
}

const _current = bowser;

export const details = {
  Chrome: {
    name: 'Chrome',
    downloadUrl: 'https://google.com/chrome',
    icon: <ChromeIcon />,
  },
  Firefox: {
    name: 'Firefox',
    downloadUrl: 'https://getfirefox.com/',
    icon: <FirefoxIcon />,
  },
  Opera: {
    name: 'Opera',
    downloadUrl: 'https://opera.com/',
    icon: <OperaIcon />,
  },
  'Microsoft Edge': {
    name: 'Edge',
    downloadUrl: 'https://www.microsoft.com/windows/microsoft-edge',
    icon: <EdgeIcon />,
    isApplicable() {
      return _current.windows && _current.osversion === '10';
    }
  },
  Safari: {
    name: 'Safari',
    downloadUrl: 'https://www.apple.com/safari/',
    icon: <SafariIcon />,
    isApplicable() {
      return _current.mac;
    }
  },
} as { [id: string]: BrowserDetails };


export const currentBrowser: Browser = {
  id: _current.name as BrowserId,
  name: details[_current.name].name,
  downloadUrl: details[_current.name].downloadUrl,
  version: _current.version,
}

export default details;