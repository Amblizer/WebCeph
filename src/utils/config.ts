const redoType: ActionType = 'REDO_REQUESTED';
const undoType: ActionType = 'UNDO_REQUESTED';

const undoableConfig = {
  undoType,
  redoType,
  limit: 100,
};

export const DEMO_IMAGE_URL =
  'https://upload.wikimedia.org/wikipedia/commons/6/66/Cephalometric_radiograph.JPG';

export { undoableConfig };

export const supportedLocales = ['en-US', 'ar-SY'];

export const bundledLocales = ['en-US'];

import zipObject from 'lodash/zipObject';
import map from 'lodash/map';

export const bundleLocaleData = zipObject(
  bundledLocales,
  map(bundledLocales, (locale) => {
    return require(`json-loader!locale/${locale}.json`) as Locale;
  }),
);

