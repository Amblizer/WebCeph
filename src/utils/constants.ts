export enum ErrorCode {
  INCOMPATIBLE_IMAGE_TYPE,
  UNKNOWN,
}

/**
 * Events are just Redux actions. 
 * The name 'event' however describes better what an action does, 
 * because Redux action should never produce side effects, they 
 * just indicate that something has happend.
 */
export const Event = {
  ADD_POINT_REQUESTED: 'ADD_POINT_REQUESTED',
  SET_ACTIVE_ANALYSIS_REQUESTED: 'SET_ACTIVE_ANALYSIS_REQUESTED',
  FLIP_IMAGE_X_REQUESTED: 'FLIP_IMAGE_X_REQUESTED',
  SET_IMAGE_BRIGHTNESS_REQUESTED: 'SET_IMAGE_BRIGHTNESS_REQUESTED',
  INVERT_IMAGE_REQUESTED: 'INVERT_IMAGE_REQUESTED',
  LOAD_IMAGE_REQUESTED: 'LOAD_IMAGE_REQUESTED',
  LOAD_IMAGE_SUCCEEDED: 'LOAD_IMAGE_SUCCEEDED',
  LOAD_IMAGE_FAILED: 'LOAD_IMAGE_FAILED',
  WORKER_CREATED: 'WORKER_CREATED',
  WORKER_TERMINATED: 'WORKER_TERMINATED',
  WORKER_STATUS_CHANGED: 'WORKER_STATUS_CHANGED',
  SET_IS_CEPHALO_REQUESTED: 'SET_IS_CEPHALO_REQUESTED',
  RESET_WORKSPACE_REQUESTED: 'RESET_WORKSPACE_REQUESTED',
  CANVAS_RESIZED: 'CANVAS_RESIZED',
};

export enum ImageWorkerAction {
  PERFORM_EDITS,
  IS_CEPHALO,
}