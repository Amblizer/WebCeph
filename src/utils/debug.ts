export function printUnexpectedPayloadWarning(type: string, state: any) {
  __DEBUG__ && console.warn(
    'Did not expect action payload to be undefined for action ' +
    type + '. ' +
    'Returning current state.',
    state,
  );
}

export const delay = (timeout: number, resolveValue: any) => {
  return new Promise((resolve) => {
    setTimeout(
      resolve.bind(null, resolveValue),
      timeout,
    );
  });
};
