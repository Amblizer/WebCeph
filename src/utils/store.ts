export function createAction<T extends ActionType>(type: T, payload: Events[T]): Action<T> {
  return {
    type,
    payload,
  };
};

export function createActionCreator<T extends ActionType>(type: T) {
  return (payload: Events[T]) => createAction<T>(type, payload);
};

export function handleActions<Key extends StoreKey>(
  map: ActionToReducerMap<Key>, defaultState: StoreState[Key]
): Reducer<StoreState[Key], ActionType> {
  return (state: StoreState[Key], action: Action<ActionType>) => {
    const reducer: Reducer<StoreState[Key], typeof action['type']> | undefined = map[action.type];
    if (typeof reducer !== 'undefined') {
      return reducer(state, action);
    }
    if (typeof state !== 'undefined') {
      return state;
    }
    return defaultState;
  };
}