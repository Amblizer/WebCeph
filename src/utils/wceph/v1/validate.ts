import {
  WCephJSON,
  JSON_FILE_NAME,
} from './format';

import JSZip from 'jszip';

import filter from 'lodash/filter';
import map from 'lodash/map';
import negate from 'lodash/negate';
import isUndefined from 'lodash/isUndefined';
import isPlainObject from 'lodash/isPlainObject';
import every from 'lodash/every';
import has from 'lodash/has';
import keys from 'lodash/keys';
import values from 'lodash/values';
import isBoolean from 'lodash/isBoolean';
import isNumber from 'lodash/isNumber';
import isString from 'lodash/isString';

import { isGeometricalObject } from 'utils/math';

const isTrue = (value: any) => value === true;
const isDefined = negate(isUndefined);

type Rule<T> = (data: T) => true | false;
type ErrorMaker<T> = (data: T) => ValidationError;
type Fixer<T> = (data: T, error: ValidationError) => T;

enum ValidationErrorType {
  UNSPECIFIED_FILE_VERSION,
  NO_REFS,
  MISSING_REFS,
  MISSING_DATA,
  INCOMPATIBLE_IMAGE_TYPE,
  INCOMPATIBLE_BRIGHTNESS_VALUE,
  INCOMPATIBLE_CONTRAST_VALUE,
  INVALID_IMAGE_DATA,
  INVALID_TRACING_DATA,
}

const getMessageForError = (type: ValidationErrorType, data?: any) => {
  switch (type) {
    case ValidationErrorType.UNSPECIFIED_FILE_VERSION:
      return (
        `Unspecified file version`
      );
    case ValidationErrorType.MISSING_REFS:
      return (
        `Invalid file format`
      );
    default:
      return ValidationErrorType[type];
  }
};

function createErrorMaker<T>(type: ValidationErrorType): ErrorMaker<T> {
  return (data: T) => ({
    type,
    message: getMessageForError(type, data),
    data,
  });
};

// @TODO: validate json structure
const rules: [
  Rule<WCephJSON>,
  ErrorMaker<WCephJSON>,
  Fixer<WCephJSON> | undefined
][] = [
  [
    ({ version }) => isDefined(version),
    createErrorMaker(ValidationErrorType.UNSPECIFIED_FILE_VERSION),
    undefined,
  ],
  [
    ({ version }) => version === 1,
    createErrorMaker(ValidationErrorType.UNSPECIFIED_FILE_VERSION),
    undefined,
  ],
  [
    ({ refs }) => (
      isPlainObject(refs) &&
      isPlainObject(refs.images) &&
      isPlainObject(refs.thumbs)
    ),
    createErrorMaker(ValidationErrorType.NO_REFS),
    undefined,
  ],
  [
    ({ data }) => isDefined(data) && isPlainObject(data),
    createErrorMaker(ValidationErrorType.MISSING_DATA),
    undefined,
  ],
  [
    ({ data, refs }) => {
      return isPlainObject(refs.images) && every(keys(data), key => has(refs.images, key));
    },
    createErrorMaker(ValidationErrorType.MISSING_REFS),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), (image) => {
        return (
          isString(image.type) &&
          isBoolean(image.flipX) &&
          isBoolean(image.flipY) &&
          isBoolean(image.invertColors) &&
          isNumber(image.brightness) &&
          isNumber(image.contrast)
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_IMAGE_DATA),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ tracing }) => {
        return (
          isPlainObject(tracing) &&
          isString(tracing.mode) &&
          (
            isNumber(tracing.scaleFactor) ||
            tracing.scaleFactor === null
          )
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_TRACING_DATA),
    undefined,
  ],

  [
    ({ data }) => {
      return every(values(data), ({ tracing: { mode } }) => {
        return (
          mode === 'auto' ||
          mode === 'assisted' ||
          mode === 'manual'
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_TRACING_DATA),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ tracing: { manualLandmarks } }) => {
        return (
          isPlainObject(manualLandmarks) &&
          every(values(manualLandmarks), isGeometricalObject)
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_TRACING_DATA),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ tracing: { skippedSteps } }) => {
        return (
          isPlainObject(skippedSteps) &&
          every(values(skippedSteps), isTrue)
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_TRACING_DATA),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ type }) => {
        return (
          type === null ||
          type === 'ceph_lateral' ||
          type === 'ceph_pa' ||
          type === 'photo_lateral' ||
          type === 'photo_frontal' ||
          type === 'panoramic'
        );
      });
    },
    createErrorMaker(ValidationErrorType.INCOMPATIBLE_IMAGE_TYPE),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ brightness }) => {
        return brightness >= 0 && brightness <= 1;
      });
    },
    createErrorMaker(ValidationErrorType.INCOMPATIBLE_BRIGHTNESS_VALUE),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ contrast }) => {
        return contrast >= 0 && contrast <= 1;
      });
    },
    createErrorMaker(ValidationErrorType.INCOMPATIBLE_CONTRAST_VALUE),
    undefined,
  ],
  [
    ({ data }) => {
      return every(values(data), ({ analysis }) => {
        return (
          isString(analysis) || analysis === null
        );
      });
    },
    createErrorMaker(ValidationErrorType.INVALID_TRACING_DATA),
    undefined,
  ],
];

export const validateIndexJSON = (json: WCephJSON): ValidationError[] => {
  return map(
    filter(
      map(rules, rule => {
        const validator = rule[0];
        const errorMaker = rule[1];
        if (validator(json) === true) {
          return true;
        }
        return errorMaker;
      }),
      result => result !== true,
    ),
    (makeError: ErrorMaker<WCephJSON>) => makeError(json),
  );
};

const validateFile: WCeph.Validator = async (fileToValidate, options) => {
  const {

  } = options;
  const zip = new JSZip();
  await zip.loadAsync(fileToValidate);
  const json: WCephJSON = JSON.parse(
    await zip.file(JSON_FILE_NAME).async('string')
  );
  return [
    // @TODO: add unzip errors
    ...validateIndexJSON(json),
  ];
};

export default validateFile;
