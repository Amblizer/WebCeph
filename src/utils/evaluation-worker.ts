declare var self: DedicatedWorkerGlobalScope;
import reject from 'lodash/reject';
import every from 'lodash/every';
import has from 'lodash/has';

import { evaluate, getStepsForAnalysis } from '../analyses/helpers';
import { addManualLandmark, tryAutomaticSteps } from '../actions/workspace';
import { Event } from '../utils/constants';

const isManual = (step: CephaloLandmark) => step.type === 'point';

import common from '../analyses/common';
const steps: CephaloLandmark[] = getStepsForAnalysis(common);
const getCephaloMapper = (mappedLandmarks: { [id: string]: GeometricalObject }) => {
  const toPoint = (cephaloPoint: CephaloPoint) => {
    return mappedLandmarks[cephaloPoint.symbol] as GeometricalPoint;
  };
  const toLine = (cephaloLine: CephaloLine) => {
    const A = toPoint(cephaloLine.components[0]);
    const B = toPoint(cephaloLine.components[1]);
    return {
      x1: A.x,
      y1: A.y,
      x2: B.x,
      y2: B.y,
    }
  };
  // @TODO: Handle scale factor
  return { toPoint, toLine, scaleFactor: 1 / 3.2 };
}

const getStepStateSelector = (stepsBeingEvaluated, mappedLandmarks, expectedLandmark) => (landmark: CephaloLandmark): StepState => {
  if (has(mappedLandmarks, landmark.symbol)) {
    return 'done';
  } else if (has(stepsBeingEvaluated, landmark.symbol)) {
    return 'evaluating';
  } else if (expectedLandmark && expectedLandmark.symbol === landmark.symbol) {
    return 'current';
  } else {
    return 'pending';
  }
};

self.addEventListener('message', ({ data }) => {
  const { mappedLandmarks, stepsBeingEvaluted, expectedLandmark } = data;
  const getStepState = getStepStateSelector(stepsBeingEvaluted, mappedLandmarks, expectedLandmark);
  const cephaloMapper: CephaloMapper = getCephaloMapper(mappedLandmarks);
  const eligibleSteps = reject(steps, s => isManual(s) || getStepState(s) !== 'pending');
  for (const step of eligibleSteps) {
    console.info('Evaluationg step %s for automatic tracing', step.symbol);
    if (every(
      step.components,
      component => getStepState(component) === 'done',
    )) {
      console.log('Found step eligible for automatic evaluation', step.symbol);
      const value = evaluate(step, cephaloMapper);
      if (value) {
        postMessage(addManualLandmark(step.symbol, value));
        postMessage({ type: Event.STEP_EVALUATION_FINISHED, payload: step.symbol });
        postMessage(tryAutomaticSteps());
      } else {
        postMessage({ type: Event.STEP_EVALUATION_FINISHED, payload: step.symbol });
      }
    } else {
      console.info('Step %s is not eligible for automatic tracing', step.symbol);
      // yield put({ type: Event.STEP_EVALUATION_FINISHED, payload: step.symbol });
      self.postMessage('END');
    }
  };
});
