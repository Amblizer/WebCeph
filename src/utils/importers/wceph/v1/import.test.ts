import fs from 'fs';
import importFile from './import';
import { isV1GeometricalPoint } from './validate';
import { isActionOfType } from 'utils/store';

test.skip('WCeph Importer', async () => {
  const buffer = fs.readFileSync(__dirname + '/fixtures/valid/1.wceph').buffer;
  const file = new File([buffer], 'Test WCeph 1.wceph');
  const actions = await importFile(file, { });
  expect(actions.length).toBeGreaterThan(0);
  for (const action of actions) {
    if (isActionOfType(action, 'SET_IMAGE_PROPS')) {
      expect(action.payload.tracing).toBeDefined();
      expect(action.payload.tracing!.manualLandmarks).toBeDefined();
      // tslint:disable-next-line no-string-literal
      const N = action.payload.tracing!.manualLandmarks!['N'];
      expect(N).toBeDefined();
      expect(isV1GeometricalPoint(N)).toBe(true);
      expect(action.payload.flipX).toBe(true);
      expect(action.payload.tracing).toMatchSnapshot();
    }
  }
});
