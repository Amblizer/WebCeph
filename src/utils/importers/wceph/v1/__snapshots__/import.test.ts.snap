exports[`test WCeph Importer 1`] = `
Object {
  "manualLandmarks": Object {
    "A": Object {
      "x": 799.8980432798951,
      "y": 1152,
    },
    "ANS": Object {
      "x": 753.8180414412841,
      "y": 1114.2981818181818,
    },
    "Ar": Object {
      "x": 1813.658083729335,
      "y": 1118.4872727272727,
    },
    "B": Object {
      "x": 804.0871343561323,
      "y": 1604.421818181818,
    },
    "Gn": Object {
      "x": 766.3853146699962,
      "y": 1872.5236363636363,
    },
    "Go": Object {
      "x": 1637.7162585273659,
      "y": 1621.1781818181817,
    },
    "L1 Apex": Object {
      "x": 862.7344094234554,
      "y": 1562.530909090909,
    },
    "L1 Incisal Edge": Object {
      "x": 799.8980432798951,
      "y": 1336.32,
    },
    "Me": Object {
      "x": 829.2216808135565,
      "y": 1885.090909090909,
    },
    "N": Object {
      "x": 900.4362291095915,
      "y": 490.12363636363636,
    },
    "Or": Object {
      "x": 1055.4325989303736,
      "y": 867.1418181818182,
    },
    "PNS": Object {
      "x": 1373.8035207244127,
      "y": 1198.08,
    },
    "Po": Object {
      "x": 1897.439905254082,
      "y": 929.9781818181817,
    },
    "Pog": Object {
      "x": 749.6289503650469,
      "y": 1780.3636363636363,
    },
    "S": Object {
      "x": 1692.1744425184515,
      "y": 737.28,
    },
    "U1 Apex": Object {
      "x": 850.1671361947433,
      "y": 1152,
    },
    "U1 Incisal Edge": Object {
      "x": 749.6289503650469,
      "y": 1382.3999999999999,
    },
  },
  "mode": "assisted",
  "scaleFactor": null,
  "skippedSteps": Object {},
}
`;
